tests/testsuite/global_cache_tracker.rs

```diff
@@ -0,0 +1,1865 @@
+    let cache_dir = paths::home().join(".cargo/registry/cache/example.com-a6c4a5adcb232b9a");
+    let src_dir = paths::home().join(".cargo/registry/src/example.com-a6c4a5adcb232b9a");
+
+    GlobalCacheTracker::db_path(&config)
+        .into_path_unlocked()
+        .rm_rf();
+
+    let _lock = config
+        .acquire_package_cache_lock(CacheLockMode::MutateExclusive)
+        .unwrap();
+    let mut tracker = GlobalCacheTracker::new(&config).unwrap();
+    let mut deferred = DeferredGlobalLastUse::new();
+
+    cache_dir.rm_rf();
+    cache_dir.mkdir_p();
+    src_dir.rm_rf();
+    src_dir.mkdir_p();
+    paths::home()
+        .join(".cargo/registry/index/example.com-a6c4a5adcb232b9a")
+        .mkdir_p();
+    let mut create = |name: &str, age, crate_size: u64, src_size: u64| {
+        let crate_filename = InternedString::new(&format!("{name}.crate"));
+        deferred.mark_registry_crate_used_stamp(
+            global_cache_tracker::RegistryCrate {
+                encoded_registry_name: "example.com-a6c4a5adcb232b9a".into(),
+                crate_filename,
+                size: crate_size,
+            },
+            Some(&days_ago(age)),
+        );
+        deferred.mark_registry_src_used_stamp(
+            global_cache_tracker::RegistrySrc {
+                encoded_registry_name: "example.com-a6c4a5adcb232b9a".into(),
+                package_dir: name.into(),
+                size: Some(src_size),
+            },
+            Some(&days_ago(age)),
+        );
+        std::fs::write(
+            cache_dir.join(crate_filename),
+            "x".repeat(crate_size as usize),
+        )
+        .unwrap();
+        let path = src_dir.join(name);
+        path.mkdir_p();
+        std::fs::write(path.join("data"), "x".repeat(src_size as usize)).unwrap()
+    };
+
+    for (name, age, crate_size, src_size) in test_crates {
+        create(name, *age, *crate_size, *src_size);
+    }
+    deferred.save(&mut tracker).unwrap();
+
+    (cache_dir, src_dir)
+}
+
+#[cargo_test]
+fn auto_gc_gated() {
+    // Requires -Zgc to both track last-use data and to run auto-gc.
+    let p = basic_foo_bar_project();
+    p.cargo("check")
+        .env("__CARGO_TEST_LAST_USE_NOW", months_ago_unix(4))
+        .run();
+    // Check that it did not create a database or delete anything.
+    let config = ConfigBuilder::new().build();
+    assert!(!GlobalCacheTracker::db_path(&config)
+        .into_path_unlocked()
+        .exists());
+    assert_eq!(get_index_names().len(), 1);
+
+    // Again in the future, shouldn't auto-gc.
+    p.cargo("check").run();
+    assert!(!GlobalCacheTracker::db_path(&config)
+        .into_path_unlocked()
+        .exists());
+    assert_eq!(get_index_names().len(), 1);
+}
+
+#[cargo_test]
+fn cache_clean_options_gated() {
+    // Checks that all cache clean options require -Zgc.
+    let p = project().build();
+    for opt in [
+        "--dry-run",
```

Comment:
Looks like the dry-run unstable test is duplicated

Reply:
Similar, this was just an issue with this being behind #12638.